Add subcommand to check the node state and perform repairing of storage

ktock opened the pull request:
<!--  Thanks for sending a pull request!

Please be aware that we're following the Kubernetes guidelines of contributing
to this project. This means that we have to use this mandatory template for all
of our pull requests.

Please also make sure you've read and understood our contributing guidelines
(https://github.com/cri-o/cri-o/blob/main/CONTRIBUTING.md) as well as ensuring
that all your commits are signed with `git commit -s`.

Here are some additional tips for you:

- If this is your first time, please read our contributor guidelines:
  https://git.k8s.io/community/contributors/guide#your-first-contribution and
  developer guide
  https://git.k8s.io/community/contributors/devel/development.md#development-guide
- Please label this pull request according to what type of issue you are
  addressing, especially if this is a release targeted pull request. For
  reference on required PR/issue labels, read here:
  https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
- If you want *faster* PR reviews, read how:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
- If the PR is unfinished, see how to mark it:
  https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Uncomment only one `/kind <>` line, hit enter to put that in a new line, and
remove leading whitespace from that line:
-->


/kind feature


#### What this PR does / why we need it:

Discussed in: https://github.com/openshift/enhancements/pull/1600#pullrequestreview-2268904850

This PR adds a subcommand discussed in [this thread](https://github.com/openshift/enhancements/pull/1600#pullrequestreview-2268904850). cc @mtrmac @haircommander @harche

```
crio check-layer-state <layer_integrity_state_marker> <layer_integrity_state_file>
```

This intends to run before crio starts as a CRI runtime.

This subcommand is used for detecting incompatible changes in storage configuration and performing repairing of the storage.

Concretely:

- When it starts for the first time on the node, it creates a file (call it "state file") at the location specified by layer_integrity_state_file with the contents of layer_integrity_state_marker (call it "state marker").
- Everytime this command starts on the node, it checks that the specified layer_integrity_state_marker exactly matches the contents of the state file at layer_integrity_state_file. If it doesn't match, it starts repairing the storage.

Caller (e.g. MCO) populate the state marker with configurations whose changes trigger repairing of the storage, e.g. `"version=1;store-format=1;ALS="estargz:/var/"`.


#### Which issue(s) this PR fixes:

<!--
Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->

<!--
Fixes #
or
None
-->
None
#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes see:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Added a subcommand to check the node state and perform repairing of the storage
```

ktock committed:
check: enable to check the node state and perform repairing of storage

Signed-off-by: Kohei Tokunaga <ktokunaga.mail@gmail.com>

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+)
+
+var CheckLayerStateCommand = &cli.Command{
+	Name: "check-layer-state",
+	Usage: `Check if state marker matches to the state flie contents.
+If they don't match, perform repairing of the storage.`,
```
```suggestion
	Name: "check-layer-state",
	Usage: `Check if the state marker matches the state file contents.
If they don't match, perform storage directory repair.`,
```

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+			}
+		}()
+
+		statemarker := c.Args().Get(0)
+		statefile := c.Args().Get(1)
+		if statefile == "" {
+			return fmt.Errorf("state file path is required")
+		}
```
What about making these into command-line switches?

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+
+		curMarker, err := os.ReadFile(statefile)
+		if err == nil {
+			if string(curMarker) == statemarker {
+				// no need to repair
+				return nil
+			}
```
A log line here would be useful.

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+			checkOptions := cstorage.CheckEverything()
+			report, err := store.Check(checkOptions)
+			if err != nil {
+				return err
```
```suggestion
				return fmt.Errorf("unable to check storage: %w", err)
```

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+			if err != nil {
+				return err
+			}
+			options := cstorage.RepairOptions{
+				RemoveContainers: true,
+			}
```
This has no other users, please fold this into the `store.Repair()`.

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+				}
+				return errs[0]
+			}
+		} else if !os.IsNotExist(err) {
+			return err
+		}
```
What about a small refactor to drop the indentation level here?

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+				return errs[0]
+			}
+		} else if !os.IsNotExist(err) {
+			return err
```
This error needs to be wrapped.

ktock replied:
Fixed

kwilczynski commented on internal/criocli/check_layer_state.go:
```diff
@@ -0,0 +1,64 @@
+			return err
+		}
+
+		return os.WriteFile(statefile, []byte(statemarker), 0600)
```
Please, wrap an error here, with a sensible message to the user.

ktock replied:
Fixed

kwilczynski commented on cmd/crio/main.go:
```diff
@@ -157,6 +157,7 @@ func main() {
 		criocli.StatusCommand,
 		criocli.VersionCommand,
 		criocli.WipeCommand,
+		criocli.CheckLayerStateCommand,
```
Keep the list sorted, if possible.

ktock replied:
Fixed

kwilczynski commented:
/ok-to-test

kwilczynski commented:
@ktock, this is something I am not sure about... even though I worked with the check+repair functionality a lot recently.

Are you confident that c/storage handles this type of images correctly? From the check+repair point of view, that is.

I haven't personally tested both features with eStargz type images. Hence my question.

However, if it does work correctly, then why a different sub-command? We already have the `crio check`.

kwilczynski commented:
I also talked about this with @mtrmac briefly while looking into something else that touched the check+repair feature.

ktock commented:
@kwilczynski 

> Are you confident that c/storage handles this type of images correctly? From the check+repair point of view, that is.
> I haven't personally tested both features with eStargz type images. Hence my question.

Thanks for the comment. I've created an integration test at https://github.com/ktock/stargz-snapshotter/blob/refs/heads/test-crio-restart/script/cri-o-restart/test-restart.sh#L37-L72 .
This performs restarting of CRI-O with changing storage configuration (enable -> disable Additional Layer Store).

- Running `check-layer-state` can avoid storage errors after restarting CRI-O: https://github.com/ktock/stargz-snapshotter/actions/runs/11063312629/job/30739238773#step:4:3922
- NOT running `check-layer-state` can result in storage errors after restarting CRI-O: https://github.com/ktock/stargz-snapshotter/actions/runs/11063388835/job/30739431999#step:4:3923

Please let me know if I'm missing something.

> However, if it does work correctly, then why a different sub-command? We already have the crio check.

Thanks for the suggestion. That sounds great. Fixed the patch and added the following flags to `check` subcommand instead of introducing a new subcommnd.

- `check-layer-state`: Check if state marker matches to the state flie contents. If they don't match, perform repairing of the storage.
- `state-marker`: State marker
- `state-file`: Path to state file

kwilczynski commented:
/hold

So I can make a review.
